boards/PDM/Src/Io/Io_Efuse_22XS4200.c

```diff
@@ -0,0 +1,198 @@
+#include "Io_Efuse_22XS4200.h"
+
+// The SPI handle for the SPI device the E-Fuses are connected to
+static SPI_HandleTypeDef *efuse_spi_handle;
+
+void Io_Efuse_22XS4200_Init(SPI_HandleTypeDef *const hspi)
+{
+    efuse_spi_handle = hspi;
+}
+
+void Io_Efuse_22XS4200_configureEfuse(struct E_Fuse *e_fuse)
+{
+    // Global config register
+    Io_Efuse_22XS4200_WriteReg(SI_GCR_ADDR, GCR_CONFIG, e_fuse);
+
+    // Channel 0 config registers
+    Io_Efuse_22XS4200_WriteReg(
+            SI_RETRY_0_ADDR, RETRY_0_CONFIG, e_fuse);
+    Io_Efuse_22XS4200_WriteReg(
+            SI_CONFR_0_ADDR, CONFR_0_CONFIG, e_fuse);
+    Io_Efuse_22XS4200_WriteReg(SI_OCR_0_ADDR, OCR_0_CONFIG, e_fuse);
+
+    // Channel 1 config registers
+    Io_Efuse_22XS4200_WriteReg(
+            SI_RETRY_1_ADDR, RETRY_1_CONFIG, e_fuse);
+    Io_Efuse_22XS4200_WriteReg(
+            SI_CONFR_1_ADDR, CONFR_1_CONFIG, e_fuse);
+    Io_Efuse_22XS4200_WriteReg(SI_OCR_1_ADDR, OCR_1_CONFIG, e_fuse);
+}
+
+void Io_Efuse_22XS4200_configure_channel_monitoring(
+        uint8_t        selection,
+        struct E_Fuse *e_fuse)
+{
+    uint16_t reg_val = 0x0000;
+
+    // Read original content of GCR Register
+    reg_val = Io_Efuse_22XS4200_ReadReg(SI_GCR_ADDR, e_fuse);
+
+    reg_val &= ~(CSNS1_EN_MASK | CSNS0_EN_MASK);
+    reg_val |= (selection & (CSNS1_EN_MASK | CSNS0_EN_MASK));
+
+    Io_Efuse_22XS4200_WriteReg(SI_GCR_ADDR, reg_val, e_fuse);
+}
+
+void Io_Efuse_22XS4200_UpdateStatus(struct E_Fuse *e_fuse)
+{
+    uint16_t status = StatusType_NoFault;
+
+    status = Io_Efuse_22XS4200_ReadReg(SO_STATR_ADDR, e_fuse);
+
+    /* Clear other bits */
+    status &= EFUSE_SO_DATA_MASK;
+
+    e_fuse->status = (StatusType_e)status;
+}
+
+void Io_Efuse_22XS4200_UpdateFaults(struct E_Fuse *e_fuse)
+{
+    uint16_t channel_0_faults = FaultType_NoFault;
+    uint16_t channel_1_faults = FaultType_NoFault;
+
+    channel_0_faults = Io_Efuse_22XS4200_ReadReg(SO_FAULTR_0_ADDR, e_fuse);
+    channel_1_faults = Io_Efuse_22XS4200_ReadReg(SO_FAULTR_1_ADDR, e_fuse);
+
+    /* Clear other bits */
+    channel_0_faults &= EFUSE_SO_DATA_MASK;
+    channel_1_faults &= EFUSE_SO_DATA_MASK;
+
+    e_fuse->faults.channel_0_faults = (FaultType_e)channel_0_faults;
+    e_fuse->faults.channel_1_faults = (FaultType_e)channel_1_faults;
+}
+
+void Io_Efuse_22XS4200_WriteReg(
+        uint8_t        register_address,
+        uint16_t       register_value,
+        struct E_Fuse *e_fuse)
+{
+    HAL_StatusTypeDef status;
+    uint16_t          command    = 0x0000U;
+    uint8_t           parity_bit = 0;
+
+    command =
+            (uint16_t)((register_address & EFUSE_ADDR_MASK) << EFUSE_ADDR_SHIFT);
+    /* place register value to be written (bits 0->8) */
+    command = (uint16_t)(command | (register_value & EFUSE_SI_DATA_MASK));
+    /* invert watchdog bit state */
+    if (e_fuse->watch_dog_state)
```

Comment:
If the watchdog timer is used, the most-significant-bit must be toggled between each SPI transfer
![image](https://user-images.githubusercontent.com/25499626/81438421-96e8aa00-9121-11ea-9e36-45d4207205c4.png)

Reply:
Is there a check for whether watchdog timer is used? It seems like in this function we always toggle the MSB, regardless of whether the watchdog timer is enabled.

Reply:
We would either have to keep track of it by having a `watchdog_is_enabled` member in the Efuse struct or you could check if the WatchDog_Disable bit is set in the GCR register. I thought it would be simpler to do the watchdog toggling regardless since it's just setting or clear a bit?

Reply:
I would prefer checking GCR register bit for correctness and avoid confusion

Reply:
On second thought checking the GCR register would not work because if the watchdog is actually enabled and you are not toggling the WDIN bit between messages (ie. when you read from the register) you would end up triggering the watchdog timeout. Toggling the WDIN even when the watchdog timer is disabled is not in-inherently wrong it's just not necessary.
To exit Fail-safe mode on board boot-up you need to:
1. toggle the WDIN bit from 1->0 (Which makes use of this part of the code)
2. a. Disable the watchdog timer
    b. Or keep serving the timer within the timeout period

Reply:
The eval-board also toggles the bit no matter if the watchdog is enabled or disabled:
```
bool XSD1_xs_read_register(uint8_t regAddr, uint16_t *regVal)
{
  uint16_t command = 0x00;
  uint16_t parity = 0;
  int i;
  bool result;
  uint16_t buffer_pos;
  /* shift bit 3 of the address (channel number) to the D13 position */
  command = (uint16_t) ((regAddr & 0x8) << 10) | (regAddr & 0x7);
  /* add address of the STATR register */
  command = (uint16_t) (command | ((XS_SI_STATR_0 & 0x7) << 10));
  /* invert watchdog bit state */
  XSD1_WDstate = !XSD1_WDstate;
  if (XSD1_WDstate) {
    command |= 0x8000U;
  }
  /* calculate parity */
  parity = XSD1_xs_Cal_bit_for_P(command);
  /* put parity bit to D14 */
  command = (uint16_t) (command | ((parity & 0x1) << 14));
  /* fill tx buffer with read-command values, same for all devices */
  for(i = 0; i < XSD1_COMM_LENGTH; i++) {
    txBuff[i] = command;
  }
```

Reply:
Oh ok can you just leave a comment saying that it's fine then. When i read the comment in the beginning i wasn't sure if that was inherently good or bad.
```
// Invert watchdog bit state (Note: It is safe to do so even if the watchdog is disabled)
```